### ISSUE TITLE
Slow performance of ILocalizationProvider.Translate<T>()

### ISSUE THREAD

**Issue by stefanolsen** [valdisiljuconoks/LocalizationProvider#341]
During a code profiling session, I found that 36% of the running time is spent on this line:
https://github.com/valdisiljuconoks/LocalizationProvider/blob/4c0d66e652f2e75affbfe88ce30e356bbe2e9f7d/common/src/DbLocalizationProvider/LocalizationProvider.cs#L237

The test scenario is a start page with various blocks. Two of these blocks are instances of the same block type, which uses the `ILocalizationProvider.Translate<T>()` method.

Is there a way to get rid of both serialization and deserialization in that method?

**Comment by hangy** [comment]
The `JsonSerializer` being recreated for each method call seems have be a way bigger impact than the (de-)serialization if the method is called repeatedly

**Comment by stefanolsen** [comment]
@hangy Looks like a good optimization.

But is there a way to fill in the properties that avoids converting to and from JSON?

**Comment by valdisiljuconoks** [comment]
noice! :) @stefanolsen there might be some options for that. json route was the lazy old man path

**Comment by stefanolsen** [comment]
I guess it could be filled in with Reflection. Sort of like how Newtonsoft.Json fills in properties on an object. Like:

1. Create an object from type parameter.
2. Iterate all string properties and fill in from the provided translations. This might require some slicing of strings, to disregard the namespace parts of the translation keys.
3. If a property is an object, repeat in a nested level.

Although, Newtonsoft.Json runs through all the JSON tokens and then fills properties. While this is the opposite way. But it is the gist of my idea.

**Comment by valdisiljuconoks** [comment]
never trust unit tests ;)

but you are awesome guys! 🤙 I never had time to look closely on perf. there are many unoptimized anomalies..

**Comment by valdisiljuconoks** [comment]
you also motivated me to work on v9 specifically focusing on bottlenecks

**Comment by stefanolsen** [comment]
You are right about the unit tests. They are not a full proof. Can we do a prerelease NuGet (marked so solutions doesn't auto-update)? Then I could test with a solution where we use regular, nested and static resource classes.

**Comment by stefanolsen** [comment]
If you do a version 9, then I would suggest to:

- Change TypeFactory to regular dependency injection. It seems to allocate too much memory.
- Optimize ResourceKeyBuilder further.
- Consider replacing the queries with regular service class methods. They are not particularly slow. But the creation of them seems slow.
- Store the result of GetAllResources as a dictionary, so it is cached like that, as it is much faster to pull specific resources by name this way.

**Comment by stefanolsen** [comment]
@valdisiljuconoks Do you want me to merge this PR into a v9 branch instead of master?
It could also be nice to look into the data layer, like test the performance of loading from SQL Server, and the caching of data. But probably in a v9 release.

**Comment by valdisiljuconoks** [comment]
I changed base already for the PR. will go through later tonight and merge all pending stuff. we can push pre-release nuget if you wanna test out.

**Comment by valdisiljuconoks** [comment]
https://github.com/valdisiljuconoks/LocalizationProvider/issues/345

**Comment by valdisiljuconoks** [comment]
@stefanolsen are you looking for testing this is opti? I'm targeting v9 to net9...

**Comment by stefanolsen** [comment]
@valdisiljuconoks Yes, I will test on Optimizely CMS 12.

**Comment by valdisiljuconoks** [comment]
that one will be still running on net8?

**Comment by stefanolsen** [comment]
@valdisiljuconoks Yes, .NET 8. I also have a project using .NET 9. So cross-compiling for both will be great in v9.

**Comment by stefanolsen** [comment]
@valdisiljuconoks Just did a code profiling session on the same site and page, with the v9 branch build. And the results were very nice.

Now the bottleneck is moved to `CachedGetAllResourcesHandler`. Apparently, it reads a lot of individual keys from cache, which incurs waiting time on a lock. It also reads `ConcurrentDictionary.Count` every time. Both of these take about the same time.

I wonder if there is a way to avoid the KnownKeys, and maybe cache the data in a different way. So that the loading on each request gets less intensive.

**Comment by valdisiljuconoks** [comment]
@stefanolsen yeah, I never liked `KnownKeys`. try another round - GetAllResources is now using dictionary

**Comment by stefanolsen** [comment]
@valdisiljuconoks Nice! Is it materialized to dictionary before or after adding to cache?

Regarding caching. Would you prefer to cache every single resource separately (one cache key for each) and a list of all resources? Sort of like how Optimizely CMS caches many things internally.

Or would you like to cache the result of conversion (like how Optimizely CMS caches read-only content instances)? This would require some changes to the data layer, so that ReflectionConverter can load a scoped set of resources from the data store. Then it would only cache the resulting objects, so they don't need to be done on every request.
I get that raw ASP.NET should be supported as well. So in this case those helpers would probably have to cache the "resulting" items or lists as well.

**Comment by valdisiljuconoks** [comment]
GetAllResources already returns a dictionary. But I haven't decided on cache heuristics.. it's next

**Comment by stefanolsen** [comment]
I am still measuring high runtimes of this part. Mostly the JSON serialization/deserialization part.

**Comment by stefanolsen** [comment]
I it possible to merge this to a minor release number? Like 8.3.0? Or build a pre-release of version 9.0?

**Comment by valdisiljuconoks** [comment]
hei, I've been out of modern life for some time :)

let me check how to proceed here..
